{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DO NOT RUN THIS NOTEBOOK--READS.TXT NOT AVAILABLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Q1**. Finding sequencing errors.\n",
    "\n",
    "Q1. Finding sequencing errors.\n",
    "\n",
    "We generated a simulated set of sequenced DNA reads of the 16 yeast chromosomes at 30-fold coverage in the file data/HW_KMER/reads.txt. Each read has a length of 75 bases. Some of the reads have sequencing errors. We would like to identify which lines in the file contains reads with sequencing errors.\n",
    "\n",
    "(Part 1) One simple way to do this is by k-mer (a DNA sequence of length k) counting. Consider the k-mers generated by a shifting window of width k for each read. If the sequence has no error, each k-mer will likely be found in multiple copies due to the 30-fold coverage. However, if there is a random sequencing error, the k-mer will likely be found in only 1 or a few copies (depends on length of k). Use k=15. Identify the lines likely to contain reads with sequencing errors as any line containing a k-mer with a count of only 1. Because we are dealing with a potentially large number of reads, write the code so that it minimizes memory usage using streams. You may use the toolz library if you think it will be helpful. (50 points)\n",
    "\n",
    "(Part 2) The actual line numbers for reads with sequencing errors is found in 'data/HW_KMER/error_lines.txt. Line numbers start with 0, not 1. Of the reads that you believe to have errors, how many are false positives? For each false positive read, identify the chromosome(s) and starting position(s) it comes from. (50 points)\n",
    "\n",
    "Chromosome sequences in FASTA format are in\n",
    "\n",
    "```\n",
    "data/HW_KMER/chr01.fas\n",
    "data/HW_KMER/chr02.fas\n",
    "data/HW_KMER/chr03.fas\n",
    "data/HW_KMER/chr04.fas\n",
    "data/HW_KMER/chr05.fas\n",
    "data/HW_KMER/chr06.fas\n",
    "data/HW_KMER/chr07.fas\n",
    "data/HW_KMER/chr08.fas\n",
    "data/HW_KMER/chr09.fas\n",
    "data/HW_KMER/chr10.fas\n",
    "data/HW_KMER/chr11.fas\n",
    "data/HW_KMER/chr12.fas\n",
    "data/HW_KMER/chr13.fas\n",
    "data/HW_KMER/chr14.fas\n",
    "data/HW_KMER/chr15.fas\n",
    "data/HW_KMER/chr16.fas\n",
    "```\n",
    "\n",
    "**Hints**: \n",
    "\n",
    "1. Check your code on a smaller version of the original data or your own simulated data before applying it to the full data set (which can take a long time).\n",
    "2. You only need to consider k-mers from sliding windows generated for a single line. Ignore k-mers that wrap across two lines.\n",
    "3. When generating k-mers, also keep a record of the line number that the k-mer was generated from. This avoids having to later search all reads to find out where a k-mer comes from, which is a very expensive operation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import itertools as it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "path = 'data/HW_KMER/reads.txt'\n",
    "\n",
    "def getReads(path):\n",
    "    \"\"\"getReads will generate every read in reads.txt\"\"\"\n",
    "    reads_data = open(path, 'r')\n",
    "    for read in reads_data:\n",
    "        yield read.strip()\n",
    "    reads_data.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def addReadToDict(readSequence, myDict, lineNumber):\n",
    "    \"\"\"addReadsToDict will divide a read into multiple 15-mers and add the 15-mer into a dictionary\"\"\"\n",
    "    start = 0\n",
    "    end = 15\n",
    "    for read in readSequence:\n",
    "        newRead = readSequence[start:end]\n",
    "        myDict.setdefault(newRead, [])\n",
    "        myDict[newRead].append(lineNumber)\n",
    "        start = start + 1\n",
    "        end = end + 1\n",
    "        if(end > 74):\n",
    "            break\n",
    "    return myDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kmer_list = {}\n",
    "i = 0\n",
    "\n",
    "for read in getReads(path):\n",
    "    addReadToDict(read, kmer_list, i)\n",
    "    i = i+1\n",
    "    if (i>4400000): # Note: Had to break code here since kernel crashes without the break statement\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "402653280"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.getsizeof(kmer_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sequence_errors = {v[0] for k, v in kmer_list.items() if len(v) == 1} # List of line numbers corresponding to reads w/ errors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "path2 = 'data/HW_KMER/error_lines.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "errors = []\n",
    "trueErrors = open(path2, 'r')\n",
    "for error in trueErrors:\n",
    "    errors.append(int(error.strip()))\n",
    "errorsSet = set(errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{2326716, 4237450}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "falsePositives = sequence_errors - errorsSet\n",
    "sequence_errors - errorsSet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**There are two false-positives at read #'s 2,326,716 and 4,237,450.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "j = 0\n",
    "target1 = \"\"\n",
    "target2 = \"\"\n",
    "for read in getReads(path):\n",
    "    if(j == 2326716):\n",
    "        target1 = read\n",
    "        j = j+1\n",
    "    elif(j == 4237450):\n",
    "        target2 = read\n",
    "        j = j+1\n",
    "    elif(j > 4237450):\n",
    "        break\n",
    "    else:\n",
    "        j = j+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first false positive is the sequence:  GGTAGTGTGGTGTGTGGGTGTGGGTGTGGATGTGGTGTGGATGTGGTGTGGGTGTGGATGTGGGTGTGGTGTGTG\n"
     ]
    }
   ],
   "source": [
    "print(\"The first false positive is the sequence: \", target1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The second false positive is the sequence:  TTTTAATTTCGGTCAGAAAGCCGGGTAAGGTATGACAGCGAGAGTAGAGGTAGATGTGAGAGAGTGTGTGGGTGT\n"
     ]
    }
   ],
   "source": [
    "print(\"The second false positive is the sequence: \", target2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False Positive 1 occurs on Chromosome  9 at starting index position  439811\n",
      "False Positive 2 occurs on Chromosome  15 at starting index position  1091210\n"
     ]
    }
   ],
   "source": [
    "# Search every chromosome to find the false positive reads\n",
    "for i in range(1,17):\n",
    "    with open('data/HW_KMER/chr%02d.fas'%i) as f:\n",
    "        sequenceText = f.read()\n",
    "        sequenceText = sequenceText.split('\\n', 1)[-1]\n",
    "        sequenceText = sequenceText.replace('\\n', \"\")\n",
    "        res1 = re.search(target1, sequenceText)\n",
    "        res2 = re.search(target2, sequenceText)\n",
    "        if res1:\n",
    "            print(\"False Positive 1 occurs on Chromosome \", i, \"at starting index position \" , res1.start() + 1)\n",
    "        if res2:\n",
    "            print(\"False Positive 2 occurs on Chromosome \", i, \"at starting index position \" , res2.start() + 1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
